llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,147 @@ getLargeGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
+      auto ExtractIntegerFromMDNodeOperand = [=](const MDNode *N,
+                                                 unsigned OpNo) -> auto {
+        Constant *C =
+            cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+        return C->getUniqueInteger().getSExtValue();
+      };
+
+      // !sycl_used_aspects is supposed to contain unique values, no duplicates
+      // are expected here
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+        Aspects.push_back(ExtractIntegerFromMDNodeOperand(MDN, I));
+      }
+      llvm::sort(Aspects);
+    }
+
+    llvm::hash_code AspectsHash =
+        llvm::hash_combine_range(Aspects.begin(), Aspects.end());
+    Hash = static_cast<unsigned>(llvm::hash_combine(AspectsHash));
+  }
+
+  std::string getName(StringRef BaseName) const {
+    if (Aspects.empty())
+      return BaseName.str() + "-no-aspects";
+
+    std::string Ret = BaseName.str() + "-aspects";
+    for (int A : Aspects) {
+      Ret += "-" + std::to_string(A);
+    }
+    return Ret;
+  }
+
+  static UsedOptionalFeatures getTombstone() {
+    UsedOptionalFeatures Ret;
+    Ret.IsTombstoneKey = true;
+    return Ret;
+  }
+
+  static UsedOptionalFeatures getEmpty() {
+    UsedOptionalFeatures Ret;
+    Ret.IsEmpty = true;
+    return Ret;
+  }
+
+private:
+  // For DenseMap:
+  llvm::hash_code Hash = {};
+  bool IsTombstoneKey = false;
+  bool IsEmpty = false;
+
+public:
+  bool operator==(const UsedOptionalFeatures &Other) const {
+    // Tombstone does not compare equal to any other item
+    if (IsTombstoneKey || Other.IsTombstoneKey)
+      return false;
+
+    if (Aspects.size() != Other.Aspects.size())
+      return false;
+
+    for (size_t I = 0, E = Aspects.size(); I != E; ++I) {
+      if (Aspects[I] != Other.Aspects[I])
+        return false;
+    }
+
+    return IsEmpty == Other.IsEmpty;
+  }
+
+  unsigned hash() const { return static_cast<unsigned>(Hash); }
+};
+
+struct UsedOptionalFeaturesAsKeyInfo {
+  static inline UsedOptionalFeatures getEmptyKey() {
+    return UsedOptionalFeatures::getEmpty();
+  }
+
+  static inline UsedOptionalFeatures getTombstoneKey() {
+    return UsedOptionalFeatures::getTombstone();
+  }
+
+  static unsigned getHashValue(const UsedOptionalFeatures &Value) {
+    return Value.hash();
+  }
+
+  static bool isEqual(const UsedOptionalFeatures &LHS,
+                      const UsedOptionalFeatures &RHS) {
+    return LHS == RHS;
+  }
+};
+} // namespace
+
+std::unique_ptr<ModuleSplitterBase>
+getSplitterByOptionalFeatures(ModuleDesc &&MD,
+                              bool EmitOnlyKernelsAsEntryPoints) {
+  EntryPointGroupVec Groups;
+
+  DenseMap<UsedOptionalFeatures, EntryPointSet, UsedOptionalFeaturesAsKeyInfo>
+      PropertiesToFunctionsMap;
+
+  Module &M = MD.getModule();
+
+  // Only process module entry points:
+  for (auto &F : M.functions()) {
+    if (!isEntryPoint(F, EmitOnlyKernelsAsEntryPoints) ||
+        !MD.isEntryPointCandidate(F)) {
```

Comment:
This "||" is confusing for an uninformed reader (like me). I would have expected to see "&&" instead, based on the methods called.

If the call is correct, can you add clarifying comment to the sources, please?

Reply:
Looking at `isEntryPointCandidate` description it seems to me that the description is correct, i.e. we might have an entry point in a module, which actually belongs to a different device image and from that point of view we need to use `||` here to avoid inclusion of that entry point to more than one module.

However, I don't quite understand  how it is possible to get a module that contains entry points which belong to another module.  That "entry point candidate" mechanism was introduced in 9a55da536b1455cfa5257ded077f2b6322332357 as a replacement for "allowed entries" introduced by 8c7bb455491749861fdef0d7f349194dd51f2549. Perhaps @kbobrovs can share some background here?